### ISSUE TITLE
Slow output of large text

### ISSUE THREAD

**Issue by hensz** [tmux/tmux#3418]
Similar to issue #849 I experience a slow output of large amount of text, e.g. by running cat on a large file. When doing the same outside of tmux with the same terminal, the output is way faster.

* tmux version (`tmux -V`): tmux 3.1c
* Platform (`uname -sp`): Linux x86_64 (SLES 15-SP3)
* $TERM inside and outside of tmux (`echo $TERM`). Outside: xterm-color, inside: screen

The effect is much more visible when using a large terminal window and a large pane. If I reduce the size of the window or even the size of the pane, the output is also significantly faster. 

When using a window and pane of approximately my display size (3840x2160) the output of a large file takes about 10s, while the out of the same file is finished in under 1s when I reduce the pane size to about a third of the screen.

**Comment by nicm** [comment]
Probably tmux thinks your terminal, whatever it is, can't keep up with the data volume and remain responsive, so it slows down.

**Comment by hensz** [comment]
I'm using Windows Terminal with WSL, but I got the same effect also with the "native" WSL shell.

**Comment by hensz** [comment]
@nicm if this is the case, can we "tell" tmux that it's assumption is wrong? As I said, out of tmux the output is way faster, just as with a smaller pane, below 1 second for my test file I mentioned earlier.

**Comment by nicm** [comment]
You can try playing with `TTY_BLOCK_START` in tty.c, but you may find that if you make it too high then tmux will not be responsive if there is fast output in one pane.

**Comment by hensz** [comment]
Sadly I can't do this. I'm using tmux on a productive system and I cannot compile a custom version.
I wonder if I'm the only one who experiences such an effect or if this is more or less "normal" behavior...

**Comment by hensz** [comment]
@nicm I just tried something else. Currently I'm using tmux on a remote machine which I connect to via ssh or et. When I use ssh and output large text, e.g. with cat, it works lightning fast, when I do the same with ssh and tmux, it's slow.
However when I run tmux locally (I'm using WSL on Windows 11) the output of large text is also fast.

So the issue is not only related to tmux, but a combination of tmux with ssh.

Does this make any difference regarding how tmux sees the terminal?

**Comment by nicm** [comment]
They have different goals. ssh is happy to fill the TCP buffers and let the terminal chew through it, the only thing it cares about responding to is ^C. tmux wants you to be able to, for example, create new windows and type in other panes.

**Comment by hensz** [comment]
I understand that, but why is tmux behaving so radically different when using it over ssh than using it locally?

**Comment by nicm** [comment]
Probably the network accepts data more slowly than the terminal so tmux slows down so it doesn't get stuck behind huge buffers. You can check the `client_discarded` format to see if tmux is discarding data and redrawing instead (`tmux display -p '#{client_discarded}'`).

If so, perhaps it is doing it too aggressively with your connection/terminal size, you can test by changing `TTY_BLOCK_START`. I'd start by changing it to something huge then doing `split-window -dh 'yes'` and see if tmux is still (fairly!) responsive after 30 seconds or so.

**Comment by nicm** [comment]
You could try setting it to SIZE_MAX so it never hits? So
```
#define TTY_BLOCK_START(tty) (SIZE_MAX)
````

**Comment by hensz** [comment]
Hi @nicm!

I tried this now, but now I'm totally confused ;-)

I don't see any change between with this change and without. But what I see now is that the output of large text is always lightning fast when I have a split window, regardless of the size of the window. I tried this with both the newest version I have (tmux next-3.4) and the one provided from my distribution, tmux 3.1c.

Only one case was slow up to now: output of large text in an existing window of a running session. But when I closed this window and created a new one in the same session, now it's also fast.

So for now, the issue seems to have resolved itself magically - but I still don't understand how and why...
